### ISSUE TITLE
`cli file.txt` should bring the zed editor into focus

### ISSUE THREAD

**Issue by jyn514** [zed-industries/zed#12557]
### Check for existing issues

- [X] Completed

### Describe the feature

if Zed is not yet running and i execute `cli file.txt`, that launches a new window that is immediately in focus. but if i then run `cli file2.txt`, that will open the file in the editor, but *not* bring focus to the editor window. it would be nice for the editor to take focus, like VSCode.

note that https://github.com/zed-industries/zed/issues/10155 is related but not the same; that's about *closing* the editor, this is about *opening* it.

**Comment by ConradIrwin** [comment]
@jyn514 I'm assuming you're on linux. What flavor?

**Comment by jyn514** [comment]
```
; cat /etc/os/release
NAME="Pop!_OS"
VERSION="22.04 LTS"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 22.04 LTS"
VERSION_ID="22.04"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=jammy
UBUNTU_CODENAME=jammy
LOGO=distributor-logo-pop-os
; uname -a
Linux pop-os 6.8.0-76060800daily20240311-generic #202403110203~1714077665~22.04~4c8e9a0 SMP PREEMPT_DYNAMIC Thu A x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by jirutka** [comment]
This should be probably possible to implement using the
[xdg-activation-v1](https://wayland.app/protocols/xdg-activation-v1) protocol.

**Comment by phromo** [comment]
I'm on linux and for me `zed <filename>` only shows `zed is already running`

```bash
$ uname -a
Linux aoeu9x586a766c 6.2.0-39-generic #40-Ubuntu SMP PREEMPT_DYNAMIC Tue Nov 14 14:18:00 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

$ zed --version
zed is already running
# note: its built from main on 2024-06-10 on e829a8c3b0564bc902cc0d0a530099be7f0f036e

$ zed ./Profile
zed is already running
```

**Comment by ConradIrwin** [comment]
@phromo The way it should work is that it listens on a unix abstract socket. If you do `ps ax` do you see another copy of zed?

If not, I'd love to know what error you're getting here: https://github.com/zed-industries/zed/blob/7c2993de755f57b2ddc6ee1acdd4f20df78dffbf/crates/zed/src/main.rs#L266. You should be able to change that code to:


```
        if let Some(error) = crate::zed::listen_for_cli_connections(open_listener.clone()) {
            dbg!(err);
            return;
        }
```

It is also possible that this would be fixed by #12756

**Comment by phromo** [comment]
oh, I see, it seems I am hitting a Linux "bug" rather than a non-feature --  sorry for hijacking the thread in that case (but my opinion about not making focus the default could still be useful).

As for the Linux behaviour, yes `zed` is running as another instance. That is `zed --help` works if no other instance is running, but whenever I have zed running I only get the `is already running` message.

The usecase I was testing was having zed running and trying to use `zed <filename>` to make zed open a file in the already-running instance.

**Comment by ConradIrwin** [comment]
@phromo The intended way this works is:

* You put the output of building the `cli` crate in your $PATH
* You put the output of building the `zed` crate in libexec where the `cli` can find it

That way the cli can manage the running daemon and running zed again will add files to it.

There's some more instructions here: https://zed.dev/docs/development/linux.

I also added a `./script/install-linux` yesterday that might be less error prone than trying to build and link your self.

**Comment by phromo** [comment]
my bad, I just checked out the source and started poking around and calling `cargo build --release` on a whim... led me to mixing up `zed` and `cli` and not reading the linux instructions properly. Thanks for your patience, and sorry for hijacking this issue.

sidenote: I took the .tar.gz from `script/bundle-linux` rather than run the full `install-linux` because I use `stow` to keep my `~/.local/` tidy (e.g. I wanted the `./configure` equivalent of `PREFIX=$HOME/.local/stow/zed-editor`)
